Add consolidated release notes for 2.7.0 (#3452)
opensearch-project/opensearch-build: release-notes/opensearch-release-notes-2.7.0.md

* Another time-saving upgrade for OpenSearch administrators comes with the availability of dynamic tenant management in OpenSearch Dashboards. Admins can view, configure, and enable or disable tenancy within Dashboards and effect those changes without a restart.
* You can now use the Performance Analyzer plugin to identify hot shards within an index, so you can take action and mitigate potential impacts on performance and availability.

### Experimental Features
OpenSearch 2.7.0 includes the following experimental features. Experimental features are disabled by default. For instructions on how to enable them, refer to the documentation for the feature.

* A new correlation engine is added to the Security Analytics toolkit, enabling a knowledge graph that can be used to identify, store, and recall connected events data to help you identify patterns and investigate relationships.
* The experimental ML Framework gains a new automatic reloading mechanism for ML models, allowing you to auto-reload deployed models when a cluster restarts or when a node rejoins the cluster.

> ### OpenSearch Ml Commons Dashboards

Reviewer: should be "ML Commons" (with a capital L). And I think the section header can say "Change to ML Commons Dashboards" or "ML Commons Dashboards Update"

> * Ml-commons introduced an API breaking change in 2.7 which changes the API response data. FE is updated accordingly. With this change, FE will no longer be compatible with OpenSearch version < 2.7 ([#154](https://github.com/opensearch-project/ml-commons-dashboards/pull/154))

Reviewer: Change "Ml-commons" to ML Commons. Also, what is FE? Should we spell it out each time it's used here?

> * Ml-commons introduced an API breaking change in 2.7 which changes the API response data. FE is updated accordingly. With this change, FE will no longer be compatible with OpenSearch version < 2.7 ([#154](https://github.com/opensearch-project/ml-commons-dashboards/pull/154))

Reviewer 2: Period missing

### OpenSearch Common Utils
* InjectSecurity - inject User object in UserInfo in threadContext. ([#396](https://github.com/opensearch-project/common-utils/pull/396))

### OpenSearch Dashboards Maps
* Add localization for opensearch base map ([#312](https://github.com/opensearch-project/dashboards-maps/pull/312))
* Add geo shape query filter ([#319](https://github.com/opensearch-project/dashboards-maps/pull/319))
* Support adding static label to document layer ([#322](https://github.com/opensearch-project/dashboards-maps/pull/322))
* Add shape filter UI button ([#329](https://github.com/opensearch-project/dashboards-maps/pull/329))
* Add tooltip to draw shape filter ([#330](https://github.com/opensearch-project/dashboards-maps/pull/330))
* Support adding field label to document layer ([#336](https://github.com/opensearch-project/dashboards-maps/pull/336))
* Add search query ability on map([#370](https://github.com/opensearch-project/dashboards-maps/pull/370))

### OpenSearch Dashboards Search Relevance

> * [Feature] Exposing Metrics for Search Comparison Tool ([#162](https://github.com/opensearch-project/dashboards-search-relevance/pull/162))

Reviewer 2: Period missing

* Add `sec_to_time` Function To OpenSearch SQL ([#1438](https://github.com/opensearch-project/sql/pull/1438))
* Add `WEEKDAY` Function to SQL Plugin ([#1440](https://github.com/opensearch-project/sql/pull/1440))
* Add `YEARWEEK` Function To OpenSearch SQL ([#1445](https://github.com/opensearch-project/sql/pull/1445))
* Add `EXTRACT` Function To OpenSearch SQL Plugin ([#1443](https://github.com/opensearch-project/sql/pull/1443))
* Add `STR_TO_DATE` Function To The SQL Plugin ([#1444](https://github.com/opensearch-project/sql/pull/1444))
* Add The `TO_SECONDS` Function To The SQL Plugin ([#1447](https://github.com/opensearch-project/sql/pull/1447))
* Added Arithmetic functions to V2 engine ([#1448](https://github.com/opensearch-project/sql/pull/1448))
* Added SIGNUM function to V2 engine ([#1442](https://github.com/opensearch-project/sql/pull/1442))
* Add `TIMESTAMPADD` Function To OpenSearch SQL Plugin ([#1453](https://github.com/opensearch-project/sql/pull/1453))
* Add `Timestampdiff` Function To OpenSearch SQL ([#1472](https://github.com/opensearch-project/sql/pull/1472))
* Add Nested Support in Select Clause (#1490) ([#1518](https://github.com/opensearch-project/sql/pull/1518))

> * Fix null response from pow/power and added missing integration testing ([#1459](https://github.com/opensearch-project/sql/pull/1459))

Reviewer 2: This is a fix, goes below

Author: Moved. Thanks!

# OpenSearch and OpenSearch Dashboards 2.7.0 Release Notes

## Release Highlights

> OpenSearch 2.7.0 includes a number of new and enhanced features for your search, observability, and analytics workloads. This release delivers production-ready functionality for segment replication, searchable snapshots, and the ability to add multiple data sources to a single dashboard—features that were previously released as experimental. Also included are new flat object fields for more efficient indexing, integrated observability features within OpenSearch Dashboards, new geospatial functionality, and updates designed to simplify administration of indexes and clusters.

Reviewer 3: ```suggestion
OpenSearch 2.7.0 includes a number of new and enhanced features for your search, observability, and analytics workloads. This release delivers production-ready functionality for segment replication, searchable snapshots, and the ability to add multiple data sources to a single dashboard — features that were previously released as experimental. Also included are new flat object fields for more efficient indexing, integrated observability features within OpenSearch Dashboards, new geospatial functionality, and updates designed to simplify administration of indexes and clusters.
```

> 

Reviewer 4: @zelinh could we add highlight here?

OpenSearch Dashboards version 2.7 has introduced an improvement where administrators can now override the NodeJs version used to run the software. This modification has the potential to affect the launch logic and may require changes to existing infrastructure in order to properly launch OpenSearch Dashboards. Further information about this change can be found at the following link: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3508

Reviewer 5: @seraphjiang Just to clarify, by default there is no impact to normal OpenSearch users but adds additional capability for administrators to override NodeJs version used to run the software (when needed?). I see that there is a caveat stating that it has potential to affect the launch logic and may require changes to existing infrastructure in order to properly launch OpenSearch Dashboards when overriding it. Can we elaborate more on that rather than pointing to a PR?

Reviewer 5: Basically, point to a documentation / readme file with the information needed for launch logic when overriding the NodeJs version used to run the software.

Reviewer 4: The change is to allow user to choose NodeJS, the implementing has issue which is addressed in https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3955

For impacted customer, they has two options, modify their infrastructure code to adapt to the change, or apply above fix as workaround before we bring to release offically.

Reviewer 4: @bbarani if we could not put in highlight, i would suggest to add 3508 to break changes. 

This break change is not breaking functionality or api interface, but impact the launch behavior, although it was for another enhancement purpose. 

I'll leave this to team to make decision is better for the best of interest for community. We found this issue in early adoption and test of 2.7.

Reviewer 5: @jhmcintyre @seraphjiang Does it make sense to add a new "Known issues/bugs" section to release notes and list this there with the workaround?

Reviewer: @bbarani @seraphjiang I would think adding a section called "Known issues" below the release highlights section would make more sense than categorizing this under "release highlights". (I don't know if I'd include "bugs" in that section title b/c this doesn't seem like a bug but open to feedback.

Reviewer 4: Thanks @bbarani  and @jhmcintyre the known issues works. 
This could help other customer to find workaround if they are facing same issue.

Reviewer: @seraphjiang can you point me to where the workaround is documented and how the admin would apply it? I'm looking at 3508 and it's not entirely clear (at least to me).

Reviewer 4: The workaround is at https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3955/ not https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3508/

The instruction is to apply the same fix in 3955, and rebuild the OpenSearch Dashboards for the artifacts. For how to build , refer to https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/DEVELOPER_GUIDE.md#building-artifacts

Reviewer: @seraphjiang @bbarani  taking a stab at writing this up. Is this accurate, and sufficiently comprehensive? 

Known Issues
 
Version 2.7.0 of OpenSearch Dashboards offers administrators the ability to manually override the version of Node.js used to run the software. If an administrator chooses this option, the change has the potential to affect the launch logic for OpenSearch Dashboards, in which case the administrator may need to modify their infrastructure code or apply a workaround to enable Dashboards to launch properly. The workaround involves making a change to the CLI shell scripts as described in this issue https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3955 and rebuilding the artifacts for Dashboards as described here https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/DEVELOPER_GUIDE.md#building-artifacts. If the administrator does not choose to override the version of Node.js used to run the software, this will not be an issue.

Reviewer 4: Overall LGTM, except `If the administrator does not choose to override the version of Node.js used to run the software, this will not be an issue.` we might remove this as it will also impact cx not override the nodejs.

  thanks @jhmcintyre

Reviewer: Thanks @seraphjiang and I'm fine removing that last line. @bbarani, does this work for you?

* The third feature moving out of experimental status is support for multiple data sources in OpenSearch Dashboards. Dynamically manage data sources across multiple OpenSearch clusters, combine visualizations into a single dashboard, and more.
* New flat object fields let you store complex JSON objects in an index without indexing all subfields separately, offering improved resource utilization.
* You can now access observability features from OpenSearch Dashboards, create your own observability dashboards, add event analytics visualizations, and more without leaving the Dashboards environment.
* Geospatial functionality receives upgrades as new shape-based filters let you filter your geospatial data by drawing a rectangle or polygon over a selected area of the map. This release also enables OpenSearch to display maps in local languages.
* OpenSearch makes managing multiple indexes simpler with the addition of component templates. These templates can help you abstract common index settings, mappings, and aliases into reusable building blocks.

> * Another time-saving upgrade for OpenSearch administrators comes with the availability of dynamic tenants management in OpenSearch Dashboards. Admins can view, configure, and enable or disable tenancy within Dashboards and effect those changes without a restart.

Reviewer: change to "dynamic tenant management"

### Experimental Features
OpenSearch 2.7.0 includes the following experimental features. Experimental features are disabled by default. For instructions on how to enable them, refer to the documentation for the feature.

* A new correlation engine is added to the Security Analytics toolkit, enabling a knowledge graph that can be used to identify, store, and recall connected events data to help you identify patterns and investigate relationships.
* The experimental ML Framework gains a new automatic reloading mechanism for ML models, allowing you to auto-reload deployed models when a cluster restarts or when a node rejoins the cluster.

## Release Details

[OpenSearch and OpenSearch Dashboards 2.7.0](https://opensearch.org/versions/opensearch-2-7-0.html) includes the following feature, enhancement, bug fixes, infrastructure, documentation, maintenance and refactoring updates.

OpenSearch [Release Notes](https://github.com/opensearch-project/OpenSearch/blob/main/release-notes/opensearch.release-notes-2.7.0.md).

OpenSearch Dashboards [Release Notes](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/release-notes/opensearch-dashboards.release-notes-2.7.0.md).

> 

Reviewer 3: Can we add a link to the documentation release notes? https://github.com/opensearch-project/documentation-website/blob/main/release-notes/opensearch-documentation-release-notes-2.7.0.md

Reviewer: I'm good with this. @zelinh ?

Author: @hdhalter @jhmcintyre Thanks for reviewing. I have added documentation release notes link below; and also updated based on reviews.

## BREAKING CHANGES

### OpenSearch Ml Commons Dashboards

> * Ml-commons introduced a API breaking change in 2.7 which changed the API response data, FE is updated accordingly. With this change, FE will no longer compatible with OpenSearch version < 2.7 ([#154](https://github.com/opensearch-project/ml-commons-dashboards/pull/154))

Reviewer 3: ```suggestion
* Ml-commons introduced an API breaking change in 2.7 which changes the API response data. FE is updated accordingly. With this change, FE will no longer be compatible with OpenSearch version < 2.7 ([#154](https://github.com/opensearch-project/ml-commons-dashboards/pull/154))
```

Reviewer: Please change "Ml" to "ML" in the subhead and copy, and spell out "ML Commons" (instead of Ml-commons) in the body copy.

## FEATURES

### OpenSearch Alerting Dashboards

> * Implemented support for configuring up to 10 data filters for query and bucket level monitors using the visual editor. ([#504](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/504))

Reviewer 3: ```suggestion
* Implemented support for configuring up to 10 data filters for query and bucket-level monitors using the visual editor. ([#504](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/504))
```

> * Implemented support for >, >=, <, and <= query operators for the doc level monitor visual editor. ([#508](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/508))

Reviewer 3: ```suggestion
* Implemented support for >, >=, <, and <= query operators for the doc-level monitor visual editor. ([#508](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/508))
```

* Add geo shape query filter ([#319](https://github.com/opensearch-project/dashboards-maps/pull/319))
* Support adding static label to document layer ([#322](https://github.com/opensearch-project/dashboards-maps/pull/322))
* Add shape filter UI button ([#329](https://github.com/opensearch-project/dashboards-maps/pull/329))
* Add tooltip to draw shape filter ([#330](https://github.com/opensearch-project/dashboards-maps/pull/330))
* Support adding field label to document layer ([#336](https://github.com/opensearch-project/dashboards-maps/pull/336))
* Add search query ability on map([#370](https://github.com/opensearch-project/dashboards-maps/pull/370))

### OpenSearch Dashboards Search Relevance
* [Feature] Exposing Metrics for Search Comparison Tool ([#162](https://github.com/opensearch-project/dashboards-search-relevance/pull/162))

### OpenSearch Index Management
* Error Prevention: Add close action. ([#728](https://github.com/opensearch-project/index-management/pull/728))
* Error Prevention: Add index priority action. ([#729](https://github.com/opensearch-project/index-management/pull/729))

> * Error Prevention: Add notification, shrink, allocation and rollup. ([#732](https://github.com/opensearch-project/index-management/pull/732))

Reviewer 3: ```suggestion
* Error Prevention: Add notification, shrink, allocation, and rollup. ([#732](https://github.com/opensearch-project/index-management/pull/732))
```

* Observability Left-Navigation now grouped at "first level" of OpenSearch Dashboards Left-Navigation fly-out [#354](https://github.com/opensearch-project/dashboards-observability/pull/354)

### OpenSearch Security Analytics
* New log types. ([#332](https://github.com/opensearch-project/security-analytics/pull/332))
* Support for multiple indices in detector input. ([#336](https://github.com/opensearch-project/security-analytics/pull/336))

### OpenSearch Security Analytics Dashboards
* Adds field mappings to edit detector pages. ([#490](https://github.com/opensearch-project/security-analytics-dashboards-plugin/pull/490))
* Correlation engine UX. ([#536](https://github.com/opensearch-project/security-analytics-dashboards-plugin/pull/536))

### OpenSearch Security
* Dynamic tenancy configurations ([#2607](https://github.com/opensearch-project/security/pull/2607))

### OpenSearch Security Dashboards Plugin
* Dynamic tenancy configurations ([#1394](https://github.com/opensearch-project/security-dashboards-plugin/pull/1394))

### OpenSearch SQL
* Create datasource API (#1458) ([#1479](https://github.com/opensearch-project/sql/pull/1479))

> * REST API for GET,PUT and DELETE ([#1502](https://github.com/opensearch-project/sql/pull/1502))

Reviewer 3: ```suggestion
* REST API for GET, PUT and DELETE ([#1502](https://github.com/opensearch-project/sql/pull/1502))
```

* Update draw filter shape ui properties ([#372](https://github.com/opensearch-project/dashboards-maps/pull/372))
* Change font opacity along with OpenSearch base map layer ([#373](https://github.com/opensearch-project/dashboards-maps/pull/373))
* Toast Warning Message if OpenSearch base map is used in conflicting regions ([#382](https://github.com/opensearch-project/dashboards-maps/pull/382))
* Add before layer id when adding documents label ([#387](https://github.com/opensearch-project/dashboards-maps/pull/387))

### OpenSearch k-NN
* Support .opensearch-knn-model index as system index with security enabled ([#827](https://github.com/opensearch-project/k-NN/pull/827))

### OpenSearch Ml Commons
* Add model auto deploy feature ([#852](https://github.com/opensearch-project/ml-commons/pull/852))
* Add memory consumption estimation for models in profile API ([#853](https://github.com/opensearch-project/ml-commons/pull/853))
* Add text docs ML input ([#830](https://github.com/opensearch-project/ml-commons/pull/830))
* Add allow custom deployment plan setting; add deploy to all nodes field in model index ([#818](https://github.com/opensearch-project/ml-commons/pull/818))

> * Add exclude nodes setting. ([#813](https://github.com/opensearch-project/ml-commons/pull/813))

Reviewer 3: ```suggestion
* Add exclude nodes setting ([#813](https://github.com/opensearch-project/ml-commons/pull/813))
```

> * set model state as partially loaded if unload model from partial nodes ([#806](https://github.com/opensearch-project/ml-commons/pull/806))

Reviewer 3: ```suggestion
* Set model state as partially loaded if unload model from partial nodes ([#806](https://github.com/opensearch-project/ml-commons/pull/806))
```

* Added Arithmetic functions to V2 engine ([#1448](https://github.com/opensearch-project/sql/pull/1448))
* Added SIGNUM function to V2 engine ([#1442](https://github.com/opensearch-project/sql/pull/1442))
* Add `TIMESTAMPADD` Function To OpenSearch SQL Plugin ([#1453](https://github.com/opensearch-project/sql/pull/1453))
* Add `Timestampdiff` Function To OpenSearch SQL ([#1472](https://github.com/opensearch-project/sql/pull/1472))
* Add Nested Support in Select Clause (#1490) ([#1518](https://github.com/opensearch-project/sql/pull/1518))
* Fix null response from pow/power and added missing integration testing ([#1459](https://github.com/opensearch-project/sql/pull/1459))

## BUG FIXES

### OpenSearch Alerting
* Issue with percolate query transforming documents with object type fields. ([#844](https://github.com/opensearch-project/alerting/issues/844))
* Notification security fix. ([#852](https://github.com/opensearch-project/alerting/pull/852))

### OpenSearch Alerting Dashboards
* Anomaly detection UI bug fixes. ([#502](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/502))

> * Fixed a bug with doc level trigger creation. ([#513](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/513))

Reviewer 3: ```suggestion
* Fixed a bug with doc-level trigger creation. ([#513](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/513))
```
